### ISSUE TITLE
Detect `not_called()` and similar mistakes in Python mocks before Python 3.12

### ISSUE THREAD

**Issue by potiuk** [astral-sh/ruff#13252]
Python 3.12 started to report errors when mock assertion is done with a "common" mistake:

```
mock.not_called()
```

This does not do what you think it does, and when you run it with Python 3.12 you will get error:

```
'not_called' is not a valid assertion. Use a spec for the mock if 'not_called' is meant to be an attribute.
```

The correct way of calling it is:

```
mock.assert_not_called()
```

Similar for other methods. It's very easy to miss it during review and for Python < 3.12 that call gets silently ignored (obviously it looks like you are calling `not_called` method on mock and it will just record the call in the mock.

For projects like Apache Airflow where we optimize most PRs and run most tests with "lowest" supported Python (Python 3.8) this leads sometimes to false-positives where the tests start to fail in Python 3.12 "canary" build where it performs full suite of tests (not mentioning that this test passing on Python 3.8 does not test anything anyway because it does not make any assertion).

It would be great to detect such "mistakes" via static checks with ruff as a new rule.

**Comment by AlexWaygood** [comment]
This might be a good addition to PGH005, assuming that that rule doesn't flag this already

**Comment by AlexWaygood** [comment]
Indeed PGH005 does not currently flag this: https://play.ruff.rs/5f47abd1-efa4-4fd9-8077-616ae57edbad.

There's a possibility of false positives here as we don't yet have type inference in Ruff. If the object is not a mock, just a regular class that has a `not_called()` method, then we'd emit a spurious error on the `not_called()` call, because we currently only have limited capability to detect the type of an object. However, this seems relatively unlikely, given that not many classes have methods with names like `not_called()`. So I think this should be fine to implement now.

**Comment by potiuk** [comment]
>  So I think this should be fine to implement now.

Would be fantastic :)

**Comment by anishgirianish** [comment]
Hi! I've opened #22830 to address this. Happy to make any adjustments if needed. Thanks!
